### PR TITLE
DT-2947: call NOMIS User Roles API from changeEmailAndRequestVerifica…

### PR DESCRIPTION
**Pull request by neilmendum**
https://dsdmoj.atlassian.net/browse/DT-2947

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 1]
```diff
@@ -110,6 +111,18 @@ class NomisUserApiService(
...
   }
 
+  fun findUserByUsername(username: String): Optional<NomisApiUserPersonDetails> {
+    if (!nomisEnabled) {
+      log.debug("Nomis integration disabled, returning empty for {}", username)
+      return Optional.empty()
+    }
+    val userDetails = webClient.get().uri("/users/{username}", username)
+      .retrieve()
+      .bodyToMono(object : ParameterizedTypeReference<NomisApiUserDetails>() {})
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:121]: you don't need to use `ParameterizedTypeReference` - that is only for `List` and other collections iirc

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 2]
```diff
@@ -110,6 +111,18 @@ class NomisUserApiService(
...
+      .block()!!
+    return Optional.of(mapUserDetailsToNomisUser(userDetails))
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:123]: don't you need to handle the 404 here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt (hunk 1 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt]: Good spot - addressed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 3]
```diff
@@ -110,6 +111,18 @@ class NomisUserApiService(
       .block()!!
   }
 
+  fun findUserByUsername(username: String): Optional<NomisApiUserPersonDetails> {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:114]: we're trying to replace all `Optional` with `null` in kotlin, so might be better to use `NomisApiUserPersonDetails?` here instead

Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt]: Gotcha. Addressed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt [hunk 4]
```diff
@@ -53,17 +53,17 @@ abstract class NomisUserService(
...
+      return verifyEmailService.changeEmailAndRequestVerification(
+        username = it.username,
+        emailInput = emailInput,
+        firstName = it.firstName,
+        fullname = it.name,
+        url = url,
+        emailType = PRIMARY
+      )
+    }
+    throw UsernameNotFoundException("Account for username $username not found")
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt:66]: you could use an elvis operator here and then convert the function to an expresson body

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt (hunk 2 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt]: Addressed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt (hunk 4 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt]: Addressed

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt [hunk 5]
```diff
@@ -49,19 +52,24 @@ abstract class NomisUserService(
...
+    getNomisUserByUsernameFromNomisUserApiService(username)?.let {
+      return verifyEmailService.changeEmailAndRequestVerification(
+        username = it.username,
+        emailInput = emailInput,
+        firstName = it.firstName,
+        fullname = it.name,
+        url = url,
+        emailType = PRIMARY
+      )
+    } ?: run {
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt:70]: don't think you need `run {}` here, should just be able do to the `throw`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt [hunk 6]
```diff
@@ -110,6 +110,34 @@ class NomisUserApiService(
...
+        log.debug("User not found in NOMIS due to {}", it.message)
+        Mono.empty()
+      }
+      .onErrorResume(WebClientResponseException::class.java) {
+        log.warn(
+          "Unable to retrieve details from NOMIS for user {} due to {}",
+          username,
+          (it as WebClientResponseException).statusCode
+        )
+        Mono.empty()
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt:133]: This doesn't look quite right - we're saying that if we're unable to call the endpoint then we will just `warn` on the exception and carry on.  Was expecting the exception still to be thrown as we can't call nomis for that user.  Not so important now, but will be later on when we add the authenticate etc.

Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/nomis/service/NomisUserApiService.kt]: Good point. Changed that by following what was done with `DeliusAuthenticationServiceException` so there's consistency between the two web clients

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt (hunk 5 above)
Comment (neilmendum) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/security/NomisUserService.kt]: Good spot, addressed

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**